internal/dbunique/unique_fields.go

```diff
@@ -0,0 +1,123 @@
+package dbunique
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/riverqueue/river/rivertype"
+	"github.com/tidwall/gjson"
+)
+
+var (
+	// uniqueFieldsCache caches the unique fields for each JobArgs type. These are
+	// global to ensure that each struct type's tags are only extracted once.
+	uniqueFieldsCache = make(map[reflect.Type][]string) //nolint:gochecknoglobals
+	cacheMutex        sync.RWMutex                      //nolint:gochecknoglobals
+)
+
+type jobArgs interface {
+	Kind() string
+}
+
+// extractUniqueValues extracts the raw JSON values of the specified keys from the JSON-encoded args.
+func extractUniqueValues(encodedArgs []byte, uniqueKeys []string) []string {
+	// Use GetManyBytes to retrieve multiple values at once
+	results := gjson.GetManyBytes(encodedArgs, uniqueKeys...)
+
+	uniqueValues := make([]string, len(results))
+	for i, res := range results {
+		if res.Exists() {
+			uniqueValues[i] = res.Raw // Use Raw to get the JSON-encoded value
+		} else {
+			// Handle missing keys as "undefined" (they'll be skipped when building
+			// the unique key). We don't want to use "null" here because the JSON may
+			// actually contain "null" as a value.
+			uniqueValues[i] = "undefined"
+		}
+	}
+
+	return uniqueValues
+}
+
+// getSortedUniqueFields uses reflection to retrieve the JSON keys of fields
+// marked with `unique:"true"`. The return values are the JSON keys using the
+// same logic as the `json` struct tag.
+func getSortedUniqueFields(args jobArgs) ([]string, error) {
+	typ := reflect.TypeOf(args)
+
+	// Handle pointer to struct
+	if typ != nil && typ.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+
+	// Ensure we're dealing with a struct
+	if typ == nil || typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("expected struct, got %T", args)
+	}
+
+	var uniqueFields []string
+
+	// Iterate over all fields
+	for i := 0; i < typ.NumField(); i++ {
+		field := typ.Field(i)
+
+		// Check for `unique:"true"` tag
+		if uniqueTag, ok := field.Tag.Lookup("unique"); ok && uniqueTag == "true" {
+			// Get the corresponding JSON key
+			jsonTag := field.Tag.Get("json")
+			if jsonTag == "" {
+				// If no JSON tag, use the field name as-is
+				uniqueFields = append(uniqueFields, field.Name)
+			} else {
+				// Handle cases like `json:"recipient,omitempty"`
+				jsonKey := parseJSONTag(jsonTag)
+				uniqueFields = append(uniqueFields, jsonKey)
+			}
+		}
+	}
+
+	// Sort the uniqueFields alphabetically for consistent ordering
+	sort.Strings(uniqueFields)
+
+	return uniqueFields, nil
+}
+
+// getSortedUniqueFieldsCached retrieves unique fields with caching to avoid
+// extracting fields from the same struct type repeatedly.
+func getSortedUniqueFieldsCached(args rivertype.JobArgs) ([]string, error) {
```

Comment:
Similar to `encoding/json`, this cache allows a given struct type's tags to only need to be parsed once per program run instead of for every struct instance.